modules/protocol/resources/META-INF/smithy/smithy4s.meta.smithy

```diff
@@ -153,36 +163,70 @@ structure typeclass {
 /// include a Service Product version of the service.
+@tags(["nocodegen"])
 @trait(selector: ":is(service)")
 structure generateServiceProduct {}
 
 /// Placing this trait on a shape will cause the generated
 /// code to have optics (Lenses or Prisms) in the companion
 /// object.
+@tags(["nocodegen"])
 @trait(selector: ":is(enum, intEnum, union, structure)")
 structure generateOptics {}
 
 /// Placing this trait on an error will cause the generated code to exclude the stacktrace
 ///  via extending scala.util.control.NoStackTrace instead of Throwable.
+@tags(["nocodegen"])
 @trait(selector: "structure :is([trait|error])")
 structure noStackTrace {}
 
 /// Allows users to manually add imports to files of generated shapes.
 /// This would be helpful when some shape needs specific import(s) in order
-/// to compile. Espically in the case you want to compose refinement types
+/// to compile. Especially in the case you want to compose refinement types
 /// and other validators.
+@tags(["nocodegen"])
 @trait
 list scalaImports {
     member: Import
 }
 
+@tags(["nocodegen"])
 @trait(
     selector: """
         :is(
             number[trait|range],
             string[trait|pattern],
             string[trait|length]
         )"""
-    conflicts: ["smithy4s.meta#unwrap"]
+    conflicts: [unwrap]
 )
 structure validateNewtype {}
+
+/// Marks the given shape to be generated in a way that allows binary-compatible evolution.
+/// For example, classes generated from such structures will not have a public copy method, but will have .withXXX methods instead.
+/// Unions will not be generated as sealed traits, but will have visitor and projection methods.
+@trait(
+    selector: ":test(structure, union)"
+    conflicts: [adt, adtMember]
+)
+@traitValidators({
+    "bincompatFriendly.NoErrors": { selector: "[trait|error]", message: "A @bincompatFriendly structure must not have the error trait." }
+    "bincompatFriendly.NoInputOutput": { selector: ":in(:root(operation :is(-[input]->, -[output]->)))", message: "A @bincompatFriendly structure must not be used as an operation input/output." }
+    "bincompatFriendly.NoAdtMemberTargets": { selector: "union > member:test(> [trait|smithy4s.meta#adtMember])", message: "Members of an @bincompatFriendly union must not target shapes that have the adtMember trait." }
+    "bincompatFriendly.NoAdtTargets": { selector: "< :in(:root(union[trait|smithy4s.meta#adt] > member))", message: "Shapes with the @bincompatFriendly trait must not be used as members of an adt union." }
+})
+structure bincompatFriendly {
+    // todo: optional parameter for rendering the auxiliary `apply` methods?
+    // like, if you enable it, you only get an `apply` with what's effectively required.
+    // that way, new optional fields and those with defaults can be added without bincompatAdded.
+    // This an be added later, if it's opt-in.
+}
+
+@trait(selector: "structure[trait|smithy4s.meta#bincompatFriendly] > member")
+@traitValidators({
```

Comment:
I see you've evolved to a higher plane of consciousness

Reply:
Only level 5 or 6 my dude